src/libraries/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs

```diff
@@ -659,121 +767,162 @@ internal void ZeroMemory()
             }
         }
 
-        internal Guid ReadGuid(int offset)
+        internal unsafe Guid ReadGuid(int offset)
         {
-            // faster than Marshal.PtrToStructure(offset, typeof(Guid))
-            byte[] buffer = new byte[16];
-            ReadBytes(offset, buffer, 0, 16);
-            return new Guid(buffer);
+            offset += BaseOffset;
+
+            Validate(offset, sizeof(Guid));
+
+            Debug.Assert(0 == offset % /* alignof(Guid) */ sizeof(int), "invalid alignment");
+
+            bool mustRelease = false;
+            try
+            {
+                DangerousAddRef(ref mustRelease);
+
+                byte* addrGuid = (byte*)DangerousGetHandle() + (uint)offset;
+
+                return new Guid(new ReadOnlySpan<byte>(addrGuid, sizeof(Guid)));
+            }
+            finally
+            {
+                if (mustRelease)
+                {
+                    DangerousRelease();
+                }
+            }
         }
-        internal void WriteGuid(int offset, Guid value)
+        internal unsafe void WriteGuid(int offset, Guid value)
         {
-            // faster than Marshal.Copy(value.GetByteArray()
-            StructureToPtr(offset, value);
+            offset += BaseOffset;
+
+            Validate(offset, sizeof(Guid));
+
+            Debug.Assert(0 == offset % /* alignof(Guid) */ sizeof(int), "invalid alignment");
+            bool mustRelease = false;
+            try
+            {
+                DangerousAddRef(ref mustRelease);
+                *((Guid*)DangerousGetHandle() + (uint)offset) = value;
+            }
+            finally
+            {
+                if (mustRelease)
+                {
+                    DangerousRelease();
+                }
+            }
         }
 
         internal DateTime ReadDate(int offset)
         {
-            short[] buffer = new short[3];
-            ReadInt16Array(offset, buffer, 0, 3);
+            Span<short> spanBuffer = stackalloc short[3];
+            ReadInt16Array(offset, spanBuffer.Slice(0, 3));
             return new DateTime(
-                unchecked((ushort)buffer[0]),   // Year
-                unchecked((ushort)buffer[1]),   // Month
-                unchecked((ushort)buffer[2]));  // Day
+                unchecked((ushort)spanBuffer[0]),   // Year
+                unchecked((ushort)spanBuffer[1]),   // Month
+                unchecked((ushort)spanBuffer[2]));  // Day
```

Comment:
This could also be rewritten as unsafe method:
```csharp
[StructLayout(LayoutKind.Sequential)]
public unsafe struct DateData
{
    public ushort Year;
    public ushort Month;
    public ushort Day;
}
DateData data = *(DateData*)ptr;
return new DateTime(data.Year, data.Month, data.Day);
```